Source/Loaders/Loader.cpp

```diff
@@ -0,0 +1,76 @@
+#include <stdio.h>
+#include <iostream>
+#include <string>
+#include <algorithm>
+#include <Loaders/Loader.h>
+
+Loader::~Loader() {}
+
+void Loader::fillBufferUntil(string end) {
+	setChunkDelimiter(end);
+
+	char c;
+
+	while (!isAtDelimiter() && !isAtEOL() && (c = nextChar()) != EOF) {
+		buffer += c;
+	}
+
+	if (c == EOF) {
+		fclose(file);
```

Comment:
So this EOF branch of the code is apparently getting called twice.

![](http://tinyimg.io/i/ANQWrYs.png)

Reply:
Really glad you took a look into this and caught the issue. Here's specifically what's happening:

1. On my machine, `da-vinci.obj` ends with an empty line. When committed to git, the empty line is stripped out and the file ends abruptly at the end of a line with face data.
2. In `ObjLoader::handleFace`, the "parser" steps through the full line, and on *my* machine stops at the new line, calls `nextLine()` one more time (bottom of the `while` loop), reaches EOF then and only then, closes the file, and breaks out of the `while` loop.
3. On your machine, `handleFace()` causes EOF to be reached, closing the file, setting the flag...and then the additional `nextLine()` tries again to advance the buffer, only to see that it's still at EOF, causing the file to be closed again.

Amusingly enough, I think the reason removing `fillBufferUntil` from `nextLine` "worked" was because `string label = readNextChunk()` was just impotently stepping through space-demarcated chunks in the lines with labels not accounted for, and eventually reaching a chunk with a matched label. Regardless, `nextLine` should do what it claims, and I should make a correction to avoid double freeing.